new-log-viewer/src/utils/data.ts

```diff
@@ -1,6 +1,87 @@
 import {Nullable} from "../typings/common";
 
 
+/**
+ * Checks if 'target' is bounded by the first and last value in a sorted array of numbers.
+ *
+ * @param data A number array sorted in ascending order.
+ * @param target
+ * @return `true` if is `target` is within bounds; `false` otherwise or when the array is empty.
+ */
+const isWithinBounds = (data: number[], target: number): boolean => {
+    const {length} = data;
+    if (0 === length) {
+        return false;
+    }
+
+    return (target >= (data[0] as number)) && (target <= (data[length - 1] as number));
+};
+
+/**
+ * Performs binary search to find the smallest index `i` in the range [0, length) where the
+ * `conditionFn` is true. Assumes that the `conditionFn` is false for some prefix of the input
+ * range and true for the remainder. The most common use is to find the index `i` of a target
+ * value in a sorted array.
+ *
+ * @param length The length of the range to search.
+ * @param conditionFn A function that takes an index and returns `true` or `false`.
+ * @return The smallest index where `conditionFn(i)` is true. If no such index exists, returns
+ * `length`.
+ * @example
+ * const arr = [1, 3, 5, 7, 10, 15, 20];
+ * const result = binarySearch(arr.length, (i) => arr[i] >= 10);
+ * console.log(result); // Output: 4 (since arr[4] is 10).
+ */
+const binarySearch = (length: number, conditionFn: (index: number) => boolean): number => {
```

Comment:
Can we rename `conditionFn` to `binaryPredicate`? I think that's usually what it's called.